dataloader/dataloader.hpp

```diff
@@ -0,0 +1,245 @@
+
+  //! Get the Test Dataset.
+  DatasetY TestY() const { return testY; }
+  //! Modify the Training Dataset.
+  DatasetY& TestY() { return testY; }
+
+  //! Get the Validation Dataset.
+  DatasetX ValidX() const { return validX; }
+  //! Modify the Validation Dataset.
+  DatasetX& ValidX() { return validX; }
+
+  //! Get the Validation Dataset.
+  DatasetY ValidY() const { return validY; }
+  //! Modify the Validation Dataset.
+  DatasetY& ValidY() { return validY; }
+
+  //! Get the Scaler.
+  ScalerType Scaler() const { return scaler; }
+  //! Modify the Sclaer.
+  ScalerType& Scaler() { return scaler; }
+
+ private:
+  /**
+   * Downloads and checks hash for given dataset.
+   */
+  void DownloadDataset(const std::string& dataset)
+  {
+    if (!Utils::PathExists(datasetMap[dataset].trainPath))
+    {
+      Utils::DownloadFile(datasetMap[dataset].trainDownloadUrl,
+          datasetMap[dataset].trainPath, dataset + "_training_data.",
+          false);
+
+      if (!Utils::CompareCRC32(datasetMap[dataset].trainPath,
+          datasetMap[dataset].trainHash))
+      {
+        mlpack::Log::Fatal << "Corrupted Training Data for " <<
+            dataset << " Downloaded." << std::endl;
+      }
+    }
+    if (!Utils::PathExists(datasetMap[dataset].testPath))
+    {
+      Utils::DownloadFile(datasetMap[dataset].trainDownloadUrl,
+          datasetMap[dataset].testPath, dataset + "_testing_data.",
+          false);
+
+      if (!Utils::CompareCRC32(datasetMap[dataset].testPath,
+          datasetMap[dataset].testHash))
+        mlpack::Log::Fatal << "Corrupted Testing Data for " <<
+            dataset << " Downloaded." << std::endl;
+    }
+  }
+
+  /**
+   * Intializes dataset map to provide access to dataset details.
+   */
+  void InitializeDatasets()
+  {
+    datasetMap.insert({"mnist", Datasets<DatasetX, DatasetY>::MNIST()});
+  }
+
+  // Utility Function to wrap indices.
+  size_t WrapIndex(int index, size_t length)
+  {
+    if (index < 0)
+      return length - size_t(std::abs(index));
+
+    return index;
+  }
+
+  //! Locally stored mappings for some well known datasets.
+  std::unordered_map<std::string,
+      DatasetDetails<DatasetX, DatasetY>> datasetMap;
+
+  //! Locally stored input for training.
+  DatasetX trainX;
+  //! Locally stored input for testing.
+  DatasetX validX;
+  //! Locally stored input for validation.
+  DatasetX testX;
+
+  //! Locally stored labels for training.
+  DatasetY trainY;
+  //! Locally stored labels for validation.
+  DatasetY validY;
+  //! Locally stored labels for testing.
+  DatasetY testY;
+
+  //! Locally Stored scaler.
+  ScalerType scaler;
+
+  //! Locally stored path of dataset.
+  std::string trainDatasetPath;
+
+  //! Locally stored path of dataset.
+  std::string testDatasetPath;
+
+  //! Locally stored ratio for train-test split.
+  double ratio;
+
+  //! Locally stored augmentation.
+  std::vector<std::string> augmentation;
+
+  //! Locally stored augmented probability.
```

Comment:
IMO `augmentation probability` would sound better.

Reply:
Agreed, Will make the changes. Thanks a lot.